+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+
+            % Terminal_lsq_cost
+            if ~isempty(model.x_ref_end_val)
+                obj.f = obj.f + h0*opts.N_finite_elements(ii)*dcs.f_lsq_T_fun(x_end,...
+                    model.x_ref_end_val,...
+                    p_global);
+            end
+
+            % Terminal constraint
+            if opts.relax_terminal_constraint_homotopy
+                error("Currently unsupported")
+            end
+            g_terminal = dcs.g_terminal_fun(x_end, z_end, v_global, p_global);
+            switch opts.relax_terminal_constraint
+              case ConstraintRelaxationMode.NONE % hard constraint
+                if opts.relax_terminal_constraint_from_above
+                    obj.g.terminal = {g_terminal, model.lbg_terminal, inf*ones(dims.n_g_terminal,1)};
+                else
+                    obj.g.terminal = {g_terminal, model.lbg_terminal, model.ubg_terminal};
+                end
+              case ConstraintRelaxationMode.ELL_1 % l_1
+                obj.w.s_terminal_ell_1 = {{'s_terminal_ell_1', dims.n_g_terminal}, 0, inf, 10};
+
+                g_terminal = [g_terminal-model.lbg_terminal-obj.w.s_terminal_ell_1();
+                    -(g_terminal-model.ubg_terminal)-obj.w.s_terminal_ell_1()];
+                obj.g.terminal = {g_terminal, -inf, 0};
+                obj.f = obj.f + obj.p.rho_terminal_p()*sum(obj.w.s_terminal_ell_1());
+              case ConstraintRelaxationMode.ELL_2 % l_2
+                     % TODO(@anton): this is as it was implemented before. should handle lb != ub?
+                obj.f = obj.f + obj.p.rho_terminal_p()*(g_terminal-model.lbg_terminal)'*(g_terminal-model.lbg_terminal);
+              case ConstraintRelaxationMode.ELL_INF % l_inf
+                obj.w.s_terminal_ell_inf = {{'s_terminal_ell_inf', 1}, 0, inf, 1e3};
+
+                g_terminal = [g_terminal-model.lbg_terminal-obj.w.s_terminal_ell_inf();
+                    -(g_terminal-model.ubg_terminal)-obj.w.s_terminal_ell_inf()];
+                obj.g.terminal = {g_terminal, -inf, 0};
+                obj.f = obj.f + obj.p.rho_terminal_p()*obj.w.s_terminal_ell_inf();
+            end
+        end
+
+        function generate_complementarity_constraints(obj)
+            import casadi.*
+            opts = obj.opts;
+            dcs = obj.dcs;
+            model = obj.model;
+            % Do Cross-Complementarity
+            % TODO(@anton) correctly handle differential rk representation.
+
+            rbp = ~opts.right_boundary_point_explicit;
+
+            v_global = obj.w.v_global();
+            p_global = obj.p.p_global();
+
+            x_prev = obj.w.x(0,0,opts.n_s);
+            z_prev = obj.w.z(0,0,opts.n_s);
+            lambda_prev = obj.w.lambda(0,0,opts.n_s);
+            c_prev = dcs.c_fun(x_prev,z_prev, v_global, [p_global;obj.p.p_time_var(1)]);
+
+            if opts.use_fesd
+                switch opts.cross_comp_mode
+                  case CrossCompMode.STAGE_STAGE
```

Comment:
in e.g. the msv example i get strange results with this cross comp mode, one might check is it bug free (maybe in simpler simulation problems)

Reply:
Yep there was a bug causing wrong cross complementarities for this mode. Fixed now.